Fixes #27457 - Add support for react router in plugins

amirfefer opened the pull request:
With this PR plugins can add routes to react-router via `addRoutes` function, this removes the need for root components in plugins.

amirfefer committed:
Fixes #27457 - Add support for react router in plugins

xprazak2 commented:
Nice. I am just curious if we can refactor somehow since the code in `ExternalRoutes` is essentially the same as we already have in [AppSwitcher](https://github.com/theforeman/foreman/blob/develop/webpack/assets/javascripts/react_app/routes/index.js#L10-L19)

sharvit commented on webpack/assets/javascripts/react_app/routes/index.js:
```diff
@@ -1,11 +1,13 @@
 
 const AppSwitcher = () => (
   <Switch>
+    <Slot id="routes" />
```
since it's only for plugins, can we call it `plugins-routes`?

sharvit commented on webpack/assets/javascripts/react_app/routes/GlobalRoute.js:
```diff
@@ -0,0 +1,22 @@
+    />
+  ));
+export const addRoutes = (id, routes) => {
+  addGlobalFill('routes', id, <ExternalRoutes routes={routes} />, 100);
```
Can you please add esdoc style comments on the `addRoutes` method?

It is not clear to me what I should supply as parameters.

sharvit reviewed (changes requested):
Thanks @amirfefer

amirfefer commented:
I've added documentation:

# Client Routing
Foreman uses `react-router` for rendering react pages without any full page reload, you can use `react-router` in your plugin as well with no boilerplate.

### Register global routes file
In the plugin registeration please add:
``` ruby
Foreman::Plugin.register :"<plugin-name>" do
  register_global_js_file 'routes'
  # some code
end
```
This tells foreman to load a `routes_index.js` file

### Register routes
Create `routes_index.js` file under `webpack` directory:

```js
import { addGlobalRoutes } from 'foremanReact/routes/ExternalRoutes';
import routes from './routes';

addGlobalRoutes('<plugin>', routes);

```

### Creating the routes
Create a `routes.js` which includes your plugin 'routes`:

```js
import React from 'react';
import IndexPage from './IndexTasks';
import ShowPage from './ShowTask';

const PluginRoutes = [
  {
    path: '/<plugin>/index',
    exact: true,
    render: props => <IndexPage {...props} />,
  },
  {
    path: '/<plugin>/:id',
    render: props => <ShowPage {...props} />,
  },
];

export default PluginRoutes;

```

xprazak2 commented:
Works well, some of the test failures seem related.

sharvit commented:
Thanks @amirfefer, any special instructions about writing tests?

amirfefer commented:
- Removing WIP
- rebased
- Tests were added

@sharvit  would you mind to have another look?

sharvit commented on webpack/assets/javascripts/react_app/routes/ForemanSwitcher.js:
```diff
@@ -0,0 +1,23 @@
+
+// map state to props
+const mapStateToProps = state => ({
+  routes: state.extendable.routes || [],
```
Am i missing something? I don't see any `routes` prop in the `ForemanSwitcher` component?

amirfefer replied:
right... forgot the condition :sweat_smile:

sharvit commented on webpack/assets/javascripts/react_app/routes/ExternalRoutes.js:
```diff
@@ -0,0 +1,23 @@
+          const railsContainer = document.getElementById('rails-app-content');
+          if (railsContainer) railsContainer.remove();
+          renderProps.location && updateLocation(renderProps.location);
+          return render(renderProps);
```
Would it make sense to create a new file call `RoutingService.js` that will handle those type of logics?
It can also contain and manage the `currentLocation` in the `routes/index.js`

amirfefer replied:
totally make sense :+1:

sharvit reviewed (changes requested):
Thanks @amirfefer

amirfefer commented:
@sharvit - Thanks!
I've added a `routingService` file that handles location as you suggested.

sharvit commented on webpack/assets/javascripts/react_app/routes/ExternalRoutes.js:
```diff
@@ -0,0 +1,22 @@
+        render={renderProps => {
+          removeRailsContent();
+          renderProps.location && updateLocation(renderProps.location);
+          return render(renderProps);
```
Could it also be added to the service? (lines 16-18)

sharvit commented on webpack/assets/javascripts/react_app/routes/ExternalRoutes.js:
```diff
@@ -0,0 +1,22 @@
+import { addGlobalFill } from '../components/common/Fill/GlobalFill';
+import { updateLocation, removeRailsContent } from './RoutingService';
+
+export const addGlobalRoutes = (id, routes) =>
```
Would it make sense to use `(id, ...routes) => ` in this case?

sharvit replied:
?

amirfefer replied:
`routes` is an array of all of the routes of a plugin, wouldn't it better than adding each route as an argument? especially when the mechanism atm to represents routes  is by array

sharvit replied:
It doesn't mean you should add each item manually, you can always use it as:
```jsx
const routes = ['some', 'items'];
const globalRoutes = addGlobalRoutes(id, ...routes);
// Or
const route = 'some-item';
const globalRoutes = addGlobalRoutes(id, route);
```

Another benefit, the `routes` will always have a default value of an empty array

amirfefer replied:
But this one
```js
const route1 = 'some-item1';
const route2 = 'some-item1';
const globalRoutes = addGlobalRoutes(id, route1, route2);
```
won't work, but the function's signature `addGlobalRoutes(id, ...routes);` suggests that it can be valid

sharvit replied:
It should work as well

sharvit commented on webpack/assets/javascripts/react_app/routes/ForemanSwitcher.js:
```diff
@@ -0,0 +1,31 @@
+  if (Object.keys(routes).length) {
+    const externalRoutes = selectComponentByWeight('routes') || {};
+
+    return <Switch>{externalRoutes.concat(children)}</Switch>;
```
I am confused, did you meant to do `externalRoutes.concat(routes)` instead?

amirfefer replied:
`externalRoutes` is the routes from plugins, while children are the routes of foreman core - ForemanSwitcher wraps our app switcher.

sharvit replied:
In this case, would it make more sense doing that logic in a selector?
```jsx
const mapStateToProps = (state, ownProps) => ({
  children: selectRoutes(state, ownProps.children),
});

// in the selector
const selectRoutes = (state, childrenRoutes) => {
  const externalRoutes = selectComponentByWeight('routes') || {};
  return externalRoutes.concat(childrenRoutes);
};
```

sharvit replied:
Don't see it implemented.

Also, it would be nice to have it using our new react-redux best-practices with a separate index.js file.

sharvit commented on webpack/assets/javascripts/react_app/routes/ForemanSwitcher.js:
```diff
@@ -0,0 +1,31 @@
+
+export const ForemanSwitcher = ({ children, routes }) => {
+  if (Object.keys(routes).length) {
+    const externalRoutes = selectComponentByWeight('routes') || {};
```
Would it make sense to put this logic in a selector and use it in `mapStateToProps` so this component will only render the routes it receives (the selector would combine them)?

sharvit commented on webpack/assets/javascripts/react_app/routes/index.js:
```diff
@@ -1,27 +1,31 @@
-          if (railsContainer) railsContainer.remove();
-          currentLocation = renderProps.location;
+          removeRailsContent();
+          updateLocation(renderProps.location);
```
IMO this whole function can live inside the routing-service.
Can be reused as well since it is similar to the render method in `ExternalRoutes.js`

sharvit commented on webpack/assets/javascripts/react_app/routes/index.js:
```diff
@@ -1,27 +1,31 @@
     ))}
     <Route
       render={child => {
+        const currentLocation = getCurrentLocation();
```
IMO the whole render method can live inside the routing-serivce

sharvit commented on webpack/stories/docs/ClientRouting.md:
```diff
@@ -0,0 +1,87 @@
+import { addGlobalRoutes } from 'foremanReact/routes/ExternalRoutes';
+import routes from './routes';
+
+addGlobalRoutes('<plugin>', routes);
```
It doesn't that clear, consider changing to: `addGlobalRoutes('my-plugin-uniqe-name', routes);`

sharvit commented on webpack/stories/docs/plugins.md:
````diff
@@ -95,3 +95,6 @@ You can make sure webpack knows about your plugin by executing script `plugin_we
 ```
+# How to extend core functionaly 
+
+You can use Slot&Fill to extend react components (See Slot&Fill section)
````
Can add a link to the Slot&Fill page?

sharvit replied:
?

sharvit reviewed (changes requested):
Thanks @amirfefer

amirfefer commented on webpack/assets/javascripts/react_app/routes/RoutingService.js:
```diff
@@ -0,0 +1,42 @@
+ * @param {Object} props - routing props
+ * @param {Function} beforeRenderingCallback - a callback to run before the the rendering, if it returns false the rendering will terminate
+ */
+export const renderRoute = (
```
this function handles location logic, plus a consumer can insert a callback to be invoked before the rendering (like the default fallback with turbolinks)
If that callback returns false it will terminate rendering

MariaAga commented:
Tested and works great in foreman tasks (while still using the foreman tasks react erb template) thanks!

sharvit commented on webpack/assets/javascripts/react_app/routes/ForemanSwitcher.js:
```diff
@@ -0,0 +1,29 @@
+export default connect(mapStateToProps)(ForemanSwitcher);
+
+ForemanSwitcher.propTypes = {
+  children: PropTypes.arrayOf(PropTypes.node).isRequired,
```
`PropTypes.node` isn't already some sort of array?

amirfefer replied:
In this case, `ForemanSwitcher` children are an array of routes

sharvit reviewed (commented):
Thanks @amirfefer, Overall looks great, just some small questions.

amirfefer commented on webpack/assets/javascripts/react_app/ReactApp/ReactApp.js:
```diff
@@ -1,13 +1,12 @@
 import PropTypes from 'prop-types';
-import { Router } from 'react-router-dom';
-import history from '../history';
+import { BrowserRouter as Router } from 'react-router-dom';
```
`BrowserRouter` handles the history object automatically

sharvit commented:
Blocked by https://github.com/theforeman/foreman/pull/7180

sharvit commented:
@amirfefer what is the status here?

MariaAga commented:
Can the `blocked by another pr` label be removed as https://github.com/theforeman/foreman/pull/7180 was merged?

amirfefer commented:
Rebased with #7236 PR and tested

mmoll commented:
test failures related :broken_heart:

mmoll commented:
hm, fails also in the turbolinks-PR...

Ron-Lavi commented:
can you rebase it on top of the connected react router PR #7186 ?

Ron-Lavi commented on webpack/assets/javascripts/react_app/components/common/Slot/SlotSelectors.js:
```diff
@@ -3,7 +3,7 @@ import { getSlotComponents } from '../../../../services/SlotsRegistry';
   getSlotComponents(slotId)
     .sort((a, b) => b.weight - a.weight)
-    .map(c => c.component);
+    .map(c => c.component) || {};
```
feels like we could use `createSelector` from `reselect` here

amirfefer replied:
the components aren't stored in redux though, but in a special registry so memorization isn't effective here.

Ron-Lavi commented on webpack/stories/docs/ClientRouting.md:
```diff
@@ -0,0 +1,87 @@
+# Client Routing
```
can this also be switched to `mdx` or `Text` component?

Ron-Lavi reviewed (commented):
Thanks @amirfefer !

xprazak2 commented:
Is this still blocked?

mmoll commented:
> Is this still blocked?

AFAICT, no. But needs a good rebase :)

amirfefer commented:
updated and rebased :)

Ron-Lavi commented on webpack/assets/javascripts/react_app/Root/ReactApp.js:
```diff
@@ -1,7 +1,7 @@
 import PropTypes from 'prop-types';
-import { Router } from 'react-router-dom';
-import history from '../history';
+import { BrowserRouter as Router } from 'react-router-dom';
```
should this PR wait for the `connected react router`? 
https://github.com/theforeman/foreman/pull/7186

amirfefer replied:
makes sense,  but the rebase is pretty the same vice versa

Ron-Lavi commented on webpack/assets/javascripts/react_app/routes/ExternalRoutes.js:
```diff
@@ -0,0 +1,25 @@
+ * @param  {String} id  plugin's id - can be its name
+ * @param  {Array}   routes an array that contains a plugin's routes
+ */
+export const addGlobalRoutes = (id, routes) =>
```
Should we rename it to `registerRoutes` to stay aligned with 
`registerReducer` and `componentRegisty.register` that we provide to plugins?

amirfefer replied:
I like it :+1:

Ron-Lavi commented on webpack/stories/docs/ClientRouting.mdx:
```diff
@@ -0,0 +1,97 @@
+import IndexPage from './IndexTasks';
+import ShowPage from './ShowTask';
+
+const ForemanTasksRoutes = [
```
Can you change all `tasks` to `plugin`

Ron-Lavi commented on webpack/stories/docs/ClientRouting.mdx:
````diff
@@ -0,0 +1,97 @@
+The plugin needs to identify that a page should be loaded via react router.
+Doing so by updating the plugin's `routes.rb`:
+```ruby
+ match '/plugin_page' => '/react#index', :via => [:get]
````
nit: remove slash in `react#index`

amirfefer replied:
the `/` renders the `foreman-core` react layout

Ron-Lavi replied:
Tested and it works nicely, didn't know that.. thanks!
how should we treat permissions?
usually they are set in the controller or in the engine file - for a controller.. 
but now that it's pointing to foreman that's an issue, I created the same file ( which contains nothing) in the plugin and it works well and might solve the permissions issue.
although the foreman react_controller also renders the layout correctly: https://github.com/theforeman/foreman/blob/develop/app/controllers/react_controller.rb#L2

so we need to do the same for plugins or solve the permissions issue.
cc @tbrisker

amirfefer replied:
when a page renders within react-router all the data is fetched via ajax requests, so permissions need special care from the UI approach, no doubt. but it isn't related just to react-router. during the host details page effort, we will try to find the best mechanism for that

Ron-Lavi reviewed (commented):
Thanks @amirfefer !
should we wait for `connect-react-router` #7186
can you add the implementation to https://github.com/theforeman/foreman_plugin_template/pull/32

amirfefer commented:
@laviro, @sharvit
rebased & updated!

sharvit commented:
Thanks @amirfefer @laviro, need another rebase after #7186 merged

Ron-Lavi commented:
@amirfefer can you rebase ? thanks!

ekohl commented:
Nit: please keep commit messages in the imperative mood: https://chris.beams.io/posts/git-commit/#imperative

tbrisker commented:
Ping @amirfefer ?

amirfefer commented:
@sharvit, @tbrisker, @laviro 
Rebased and tested with foreman plugin template

sharvit commented on webpack/assets/javascripts/react_app/routes/ExternalRoutes.js:
```diff
@@ -0,0 +1,25 @@
+ * @param  {String} id  plugin's id - can be its name
+ * @param  {Array}   routes an array that contains a plugin's routes
+ */
+export const registerRoutes = (id, routes) =>
```
I find the name `ExternalRoutes.js` a bit confusing, would it make sense to change the terminology from `external routes` to `plugin routes`?

I think it will make the developer's life easier when they want to div deep into the code to understand how plugin routes work.

sharvit commented on webpack/stories/docs/ClientRouting.mdx:
```diff
@@ -0,0 +1,97 @@
+<Meta
+  title="Introduction|Client Routing"
+  parameters={{
+    storyWeight: 120,
```
A story with a `storyWeight` of 120 already exists: https://github.com/theforeman/foreman/blob/e7c0ad82a2f327e105f3ceae62b2b6721783ae5a/webpack/stories/docs/creating-a-form.stories.mdx

Please pick a unique story-weight to represent the story order in the stories list.

sharvit commented on webpack/assets/javascripts/react_app/routes/ForemanSwitcher.js:
```diff
@@ -0,0 +1,29 @@
+});
+
+// export connected component
+export default connect(mapStateToProps)(ForemanSwitcher);
```
Can you please explain why the component should be connected?
I don't find any usage of sate/actions

amirfefer replied:
when a plugin's route is registered, the store updates as well.
connect  or `useSelector` observe the store for changes and updates the component when an action is dispatched.
I refactored it to use hooks instead of `connect`

sharvit reviewed (dismissed):
Thanks @amirfefer 👍

Ron-Lavi commented on webpack/assets/javascripts/react_app/routes/ExternalRoutes.js:
```diff
@@ -0,0 +1,25 @@
+ * @param  {Array}   routes an array that contains a plugin's routes
+ */
+export const registerRoutes = (id, routes) =>
+  routes.map(({ render, path, beforeRendering = null, ...routeProps }, index) =>
```
getting the error: `Uncaught TypeError: beforeRenderingCallback is not a function`
can you remove the `null` here?

```suggestion
  routes.map(({ render, path, beforeRendering, ...routeProps }, index) =>
```

Ron-Lavi commented on webpack/assets/javascripts/react_app/routes/RoutingService.js:
```diff
@@ -0,0 +1,42 @@
+ * @param {Object} props - routing props
+ * @param {Function} beforeRenderingCallback - a callback to run before the rendering, if it returns false the rendering will terminate
+ */
+export const renderRoute = (
+  renderFn,
+  props,
+  beforeRenderingCallback = () => true
+) => {
```
can you pass the args in an object? it's a better practice for future changes.

amirfefer replied:
this is a helper function with jsdoc documentation for the arguments, this is how all our other helper functions look like, I don't see a benefit for object here.

Ron-Lavi commented on webpack/assets/javascripts/react_app/routes/RoutingService.js:
```diff
@@ -0,0 +1,42 @@
+export const renderRoute = (
+  renderFn,
+  props,
+  beforeRenderingCallback = () => true
```
can you use `noop` here instead?

amirfefer replied:
I tried but it doesn't work in this case
`Function.Prototype` doesn't fit when `beforeRenderingCallback` is invoked

Ron-Lavi reviewed (changes requested):
Tested and it works nicely! moving from foreman router pages to plugin page became very fast :)
Left some inline requests, thanks @amirfefer

ezr-ondrej commented:
@amirfefer what is the status here? :)

ofedoren commented:
@amirfefer, needs rebase again :(

amirfefer commented:
Rebased again + fixed some tests 
You can test this with the [plugin template PR](https://github.com/theforeman/foreman_plugin_template/pull/32) 
@sharvit , @laviro , @ezr-ondrej, @ofedoren could you please re-review?

edit: all green :green_circle: a great opportunity :)

ezr-ondrej commented on webpack/assets/javascripts/react_app/routes/__test__/ForemanSwitcher.test.js:
```diff
@@ -0,0 +1,19 @@
+  routes: [<Route key="1" path="/path1" />, <Route key="2" path="path2" />],
+};
+
+const pluginRoutes = [<Route key="3" path="/externalPath" />];
```
This doesn't seem used.

ezr-ondrej commented on webpack/assets/javascripts/react_app/routes/__test__/PluginRoutes.test.js:
```diff
@@ -0,0 +1,18 @@
+  describe('rendering', () => {
+    it('Adding global routes', () => {
+      registerRoutes('some-id', routes);
+      expect(store.getState().extendable).toMatchSnapshot();
```
Why don't we use mocked store here?

ezr-ondrej commented on webpack/assets/javascripts/react_app/routes/PluginRoutes.js:
```diff
@@ -0,0 +1,26 @@
+ * @param  {String} id  plugin's id - can be its name
+ * @param  {Array}   routes an array that contains a plugin's routes
+ */
+export const registerRoutes = (id, routes) =>
```
Wouldn't this be best to live alongside the `RoutingService` in the same file?
IMO this doesn't deserver special file.

amirfefer replied:
`RoutingService` is an internal helper (unrelated functions for plugins are  exported, which might cause some confusion)  that's why I separated between the two

ezr-ondrej commented on webpack/assets/javascripts/react_app/routes/RouterSelector.js:
```diff
@@ -1,6 +1,10 @@
 export const selectRouterHash = state => selectRouterLocation(state).hash;
 export const selectLastHistoryAction = state => selectRouter(state).action;
+export const selectRoutes = coreRoutes =>
+  selectComponentByWeight('routes').concat(coreRoutes);
```
This can lead to accidental override of the core Route from plugin. Is it intended?

amirfefer replied:
I'm not sure how? afaik `concat` is like a union operation between two arrays, it does not remove duplicates

amirfefer replied:
I've switched between the two, though as I mentioned here,  the routes order has much less impact  while using `exact` routes

ezr-ondrej replied:
Unfortunatelly that broke things as the core 'default-route' needs to be last.

ezr-ondrej commented on webpack/assets/javascripts/react_app/routes/index.js:
```diff
@@ -1,53 +1,26 @@
+
+const AppSwitcher = () => (
+  <ForemanSwitch>
+    {routes.map(({ render, path, ...routeProps }) => (
+      <Route
+        path={path}
+        key={path}
+        {...routeProps}
+        render={renderProps => renderRoute(render, renderProps)}
+      />
+    ))}
```
Why don't we register the core routes to the store instead?

amirfefer replied:
The registering adds an expansive abstraction layer (including a full redux action-reducer cycle and storing  components in a special registry) looks a bit overkill to me

ezr-ondrej commented on webpack/assets/javascripts/react_app/routes/PluginRoutes.js:
```diff
@@ -0,0 +1,26 @@
+export const registerRoutes = (id, routes) =>
+  routes.map(
+    ({ render, path, beforeRendering = undefined, ...routeProps }, index) =>
+      addGlobalFill(
```
can the plugin manipulate the weight?
I'd imagine we could have core routes with weight `> 10000` and plugins could register by default with lower weight, but would be allowed to override the core route by defining the weight higher for the specific route.

amirfefer replied:
`addGlobalFill` doesn't get any weight argument, so the registered routes cannot change their location.
Route component also has an `exact` prop which makes the routes order redundant (the switcher will pick the exact route regardless of the ordering), IMHO allowing routes ordering can give us a lot of issues between plugins (also an area that we don't have good coverage), using `exact` routes should avoid this from happening.

ezr-ondrej replied:
make sense :+1: once we will have some duplicate entries, we can think about it.
Although duplicate entries could be useful for plugins in some cases :shrug:

ezr-ondrej reviewed (commented):
I've got few talking points, in general it works nice, great work :+1:
There are few issue I see though, I've left them inline as talking points, but to summarize:
1. Why are the core routes exception and are not in the redux?
2. Should we allow weights for the routes? `Optional`
3. Why is the plugin supposed to require from `PluginRoutes` and not from `RoutingService`? Plugins require registerReducer from `MountingService` I don't see reason to split it here from the Service.
4. Testing would deserve some improvements - some snapshots seems quite empty and not testing much

Main point is the require point, others can be improved over time and I'd not block it on them.

amirfefer commented:
@ezr-ondrej thanks for your review :+1: 
 unfortunately, I won't be able to refactor tests until the code freeze. As a new infrastructure feature without any usage, I think adding more tests in a follow-up PR would work.

amirfefer commented:
@ezr-ondrej thanks for your review :+1: 
hopefully, I addressed most of your concerns. 
I agree that this feature can have better test coverage, but I won't be able to refactor tests until starting the stabilization process,

amirfefer commented on webpack/assets/javascripts/react_app/routes/ForemanSwitcher/ForemanSwitcher.js:
```diff
@@ -0,0 +1,20 @@
+const ForemanSwitcher = ({ routes }) => (
+  <Switch>
+    {routes}
+    <Route
```
default route

amirfefer commented:
moved the default route to `ForemanSwitcher` so the fallback route is after core and plugins routes

ezr-ondrej commented:
The snapshots needs updating

amirfefer commented:
updated

Ron-Lavi reviewed (approved):
Tested and it works great !

ezr-ondrej commented:
I doesn't load the plugin's component css for me for some reason :thinking:

amirfefer commented:
@ezr-ondrej loading plugin's css files globally is something we need to solve (can be the same solution as global js files)
but for now @laviro created a [workaround](https://github.com/theforeman/foreman_rh_cloud/pull/358/files#diff-e91d1eb0e4435dace7ecff47e398fbcd33553853dac9f2f128f019c5db886737R17-R20), another workaround is to use js styling object instead. it's another scope though.

ezr-ondrej reviewed (approved):
Apart of the css works as expected ! :+1:
Thanks @amirfefer

ofedoren commented:
@ezr-ondrej, shouldn't we revert this? I've noticed that this has a side effect on models/new page?
Before this:
![ScreenShot-1603913966732](https://user-images.githubusercontent.com/32508194/97487981-edf76b00-195d-11eb-9081-8cea340386d0.png)
After:
![ScreenShot-1603914046272](https://user-images.githubusercontent.com/32508194/97487992-f2238880-195d-11eb-8f94-70c1ec53eb4f.png)

@amirfefer, can you take a look?

MariaAga commented:
@amirfefer  there was an issue with the cancel button not showing up, maybe it's related?
[Fixes #30774 - Show cancel btn on new hw model page](https://github.com/theforeman/foreman/pull/7957)

amirfefer commented:
Thanks, @ofedoren , @MariaAga 
Looking into it

ezr-ondrej commented:
Thanks @ofedoren and @MariaAga ! Fixed by #8114.
